mono/metadata/superluminal.c

```diff
@@ -0,0 +1,329 @@
+}
+
+struct JITEnumerationData {
+	int mNumDomains;
+	int mNumAssemblies;
+	int mNumMethods;
+};
+
+static void
+method_jit_done (MonoProfiler *prof, MonoMethod *method, MonoJitInfo *jinfo)
+{
+	method_load (mono_domain_get (), method, jinfo, FALSE);
+}
+
+static void
+on_enumerate_assembly (MonoAssembly *assembly, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumAssemblies++;
+
+	MonoImage *image = mono_assembly_get_image (assembly);
+	image_event (image, TRUE, FALSE);
+}
+
+static void
+on_enumerate_jit_method (MonoDomain *domain, MonoMethod *method, MonoJitInfo *jinfo, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumMethods++;
+
+	method_load (domain, method, jinfo, TRUE);
+}
+
+static void
+on_enumerate_domain (MonoDomain *domain, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumDomains++;
+
+	// Iterate through each assembly
+	mono_domain_assembly_foreach (domain, on_enumerate_assembly, enumerationData);
+
+	// Iterate through each JIT'ed method
+	mono_domain_jit_foreach (domain, on_enumerate_jit_method, enumerationData);
+}
+
+static void
+on_attach ()
+{
+	SUPERLUMINAL_LOG ("Enumerating JIT data...");
+
+	struct JITEnumerationData enumerationData;
+	memset (&enumerationData, 0, sizeof (enumerationData));
+
+	// Iterate through each domain
+	mono_domain_foreach (on_enumerate_domain, &enumerationData);
+
+	SUPERLUMINAL_LOG_ARGS ("Finished enumerating JIT data. Found %d domains, %d assemblies, %d methods", enumerationData.mNumDomains, enumerationData.mNumAssemblies, enumerationData.mNumMethods);
+}
+
+// This callback is called by the ETW system when tracing is started / stopped. We use it to enumerate & output information about JIT compilation that happened *before* tracing started.
+DECLSPEC_NOINLINE __inline VOID __stdcall Private_EventControlCallback (_In_ LPCGUID SourceId, _In_ ULONG ControlCode, _In_ UCHAR Level, _In_ ULONGLONG MatchAnyKeyword, _In_ ULONGLONG MatchAllKeyword, _In_opt_ PEVENT_FILTER_DESCRIPTOR FilterData, _Inout_opt_ PVOID CallbackContext)
+{
+	SUPERLUMINAL_LOG_ARGS ("EventControlCallback (%d)", ControlCode);
+
+	switch (ControlCode) {
+	case EVENT_CONTROL_CODE_ENABLE_PROVIDER: {
+		gboolean isRegular = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context);
+		gboolean isRundown = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context);
+		SUPERLUMINAL_LOG_ARGS("EventControlCallback -- IsInitialized: %s, IsRuntime: %s, IsRundown: %s", is_initialized ? "true" : "false", isRegular ? "true" : "false", isRundown ? "true" : "false");
+		if (is_initialized && CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context) {
+			on_attach ();
+		}
+	} break;
+	default:
+		break;
+	};
+}
+
+/* the entry point */
+MONO_API void
```

Comment:
@rovarma: Here is one other difference, I add `MONO_API` here so that this function is exported and can be loaded from the Mono profiler integration. In the external integration, this was exported via a .def file.

@joncham @UnityAlex: Is this the proper way to handle this?